src/ListView.vala

```diff
@@ -144,21 +186,73 @@ public class Tasks.ListView : Gtk.Grid {
 
     [CCode (instance_pos = -1)]
     private int sort_function (Gtk.ListBoxRow row1, Gtk.ListBoxRow row2) {
-        var row1_completed = ((Tasks.TaskRow) row1).completed;
-        var row2_completed = ((Tasks.TaskRow) row2).completed;
+        var row_a = (Tasks.TaskRow) row1;
+        var row_b = (Tasks.TaskRow) row2;
+
+        if (row_a.completed == row_b.completed) {
+            unowned ICal.Component comp_a = row_a.task.get_icalcomponent ();
+            unowned ICal.Component comp_b = row_b.task.get_icalcomponent ();
 
-        if (row1_completed && !row2_completed) {
+            ICal.Time start_a = comp_a.get_dtstart ();
+            ICal.Time stamp_a = comp_a.get_dtstamp ();
+
+            ICal.Time start_b = comp_b.get_dtstart ();
+            ICal.Time stamp_b = comp_b.get_dtstamp ();
+
+            if ( start_a.is_null_time () && start_b.is_null_time () ) {
+                return stamp_b.compare (stamp_a);
+
+            } else if (start_a.is_null_time () && !start_b.is_null_time ()) {
+                return 1;
+
+            } else if (start_b.is_null_time () && !start_a.is_null_time ()) {
+                return -1;
+
+            } else {
+                return start_a.compare (start_b);
+            }
+
+        } else if (row_a.completed && !row_b.completed) {
             return 1;
-        } else if (row2_completed && !row1_completed) {
+
+        } else if (row_b.completed && !row_a.completed) {
             return -1;
         }
 
         return 0;
     }
 
-    private void on_tasks_added (Gee.Collection<ECal.Component> tasks) {
+    private void header_function (Gtk.ListBoxRow lbrow, Gtk.ListBoxRow? lbbefore) {
+        if (source != null || !(lbrow is Tasks.TaskRow)) {
+            return;
+        }
+        var row = (Tasks.TaskRow) lbrow;
+        unowned ICal.Component comp = row.task.get_icalcomponent ();
+
+        if (comp.get_due ().is_null_time ()) {
+            return;
+        }
+
+        if (lbbefore != null) {
+            var before = (Tasks.TaskRow) lbbefore;
+            unowned ICal.Component comp_before = before.task.get_icalcomponent ();
+
+            if (comp_before.get_due ().compare_date_only (comp.get_due ()) == 0) {
+                return;
+            }
+        }
+
+        var due_date_time = Util.ical_to_date_time (comp.get_due ());
+        var header_label = new Granite.HeaderLabel (Tasks.Util.get_relative_date (due_date_time));
+        header_label.ellipsize = Pango.EllipsizeMode.MIDDLE;
+        header_label.margin_start = 6;
+
+        row.set_header (header_label);
+    }
+
+    private void on_tasks_added (Gee.Collection<ECal.Component> tasks, E.Source source) {
         tasks.foreach ((task) => {
-            var task_row = new Tasks.TaskRow.for_component (task, source);
+            var task_row = new Tasks.TaskRow.for_component (task, source, "@colorAccent", this.source == null);
```

Comment:
Setting the color this way is super overkill, and in the case of the check button it actually does nothing since it is already using accent color by default

Reply:
Dropped tint_color property in TaskRow.